src/libcore/ffi.rs

```diff
@@ -13,24 +13,27 @@ use ::fmt;
 /// and `*mut c_void` is equivalent to C's `void*`. That said, this is
 /// *not* the same as C's `void` return type, which is Rust's `()` type.
 ///
-/// Ideally, this type would be equivalent to [`!`], but currently it may
-/// be more ideal to use `c_void` for FFI purposes.
+/// In FFI, until `extern type` is stabilized, it is recommended to use
+/// a newtype wrapper around an empty byte array instead of a newtype wrapper
+/// around `c_void`. See the [Nomicon] for details.
```

Comment:
That pattern is specifically for opaque types.  That's not the same as `void*`, in my understanding.  I am talking about the pattern where a C header says
```C
struct Foo;

Foo *make_foo();
void do_something_with_foo(Foo *x, int y);
```
Notice the total lack of `void`.

So, maybe start the paragraph with something like "To model pointers to opaque types in FFI, until `extern type` ... empty byte array."  I wouldn't even mention newtype wrappers around `c_void`; TBH, given that `void*` does not appear on the C side, that doesn't seem like an obvious idea at all to me.  Or am I missing something?

Reply:
Thanks, updated!

> TBH, given that `void*` does not appear on the C side, that doesn't seem like an obvious idea at all to me. Or am I missing something?

Maybe that's just me then :) `void *` is often used for some arbitrary pointer to an unknown/unspecified type, so it seemed like the obvious choice to use `*mut c_void` in Rust for opaque pointers.

Reply:
This whole discussion is not about when `void *` is used in C. It’s about when `Foo *` with `struct Foo;` is used in C. Opaque types, but still different types so that a pointer for a different type can’t but used accidentally without casting.